samd: Extension for the existing port.

robert-hh opened the pull request:
These are the first two commits restarting PR #8533.  The commits are:

- Remove partially functioning support for REPL duplicated at the UART port. UART support will be added later again with full API coverage.
- Replace pins.c and pins.h in the board directories by pins.csv. pins.csv is easier to maintain.

robert-hh committed:
samd: Remove the provisional support for REPL on UART.

It was only partially working and will be rpelaced later by a full
machine.UART class implementation.

robert-hh committed:
samd: Replace pins.c and pins.h by pins.csv.

The files pins.c and pins.h are generated during the build process
using a make_pins.py script.

robert-hh commented:
@dpgeorge I suggest the following sequence for commits, which at the end will have the same state as my actual development branch. Each line would be a separate commit with clear focus.
1.  Remove the preliminary REPL on UART support. (PR done)
2.  Replace pins.c and pins.h by pins.csv and make_pins.py. (PR done)
3.  Full clock init, including µs-counter init.
4.  Update mphalport.c and mphalport.h.
5.  Implement Pin Open Drain.
6.  Add SoftI2c and SoftSPI.
7.  Add machine UART consisting of  
    - Add the pin_mux table and methods.
    - Rework the ISR table.
    - Add machine.UART and UART ISR.
    - Add uos.dupterm().
8.  Add  ADAFRUIT_ITSYBITSY_M0_EXPRESS and ADAFRUIT_FEATHER_M4_EXPRESS
9.  Extensions of the micropython set-up
    - Allow for a board specific manifest.py
    - Use the common way of executing boot.py and main.py.
    - Enable a few more MicroPython features.
10. Add the class machine.ADC.
11. Add the class machine.PWM
12. Add the machine.Timer class based on softtimer.
13. Add pin.irq()
14. Small additions to the class machine
     - Enable machine.freq(value) for SAMD51x.
    - Enable machine.disable_irq(), .enable_irq() and .idle().
15. Add the machine.WDT class to machine
16. Add uos.urandom() for SAMD51.
17. Add the machine.SPI class.
18. Add a pin_info() method to the module samd.
19. Shrink/Drop the machine.LED class. (optional, to save flash space)

robert-hh committed:
samd: Move MCU-Specific settings into a mpconfig_samdXX.h file.

Located at the boards directory. That way, the mpconfigboard.h files
are almost empty, just setting the board name and the MCU name.

robert-hh committed:
samd: Add ADAFRUIT_FEATHER_M4_EXPRESS and ADAFRUIT_ITSYBITSY_M0_EXPRESS.

I use these two boards as well for testing. Therefore is is favorable
to have them added early. Test set:

ADAFRUIT_FEATHER_M4_EXPRESS: SAMD51 with 32kHz crystal.
ADAFRUIT_ITSYBITSY_M0_EXPRESS: SAMD21 without crystal.
ADAFRUIT_ITSYBITSY_M4_EXPRESS: SAMD51 without crystal.
SEEED_XIAO: SAM21 with 32kHz crystal

robert-hh committed:
samd: Sort the source code files in Makefile.

robert-hh committed:
samd: Set up the clock configuration.

Clock settings:

GCLK0: 48 MHz (SAMD21) or 120 MHz(SAMD51)
GCLK1: 32768 Hz for driving the PLL
GCLK2: 48 MHz for tzhe peripheral clock.
GCLK3: 1 MHz (SAMD21) or 8 MHz (SAMD51) for the µs ticks timer.
GCLK8: 1 kHz for WDT (SAMD21 only)

If a 32 kHz Crystal is present, it will be used as clock source.
Otherwise the DFLL48M in open loop mode is used.

GCLK0 for SAM51 can be changed between 48 MHz and 200 MHz. The specified
range is 96 MHz - 120 MHz.

robert-hh committed:
samd: Allow changing the CPU freq with machine.freq(f).

SAMD51 only.
Accepted values are 48_000_000 to 200_000_000. The range specified
by Atmel is 96_000_000 to 120_000_000.

robert-hh committed:
samd: Extend mphalport.c and mphalport.h.

changes:
- Adding the interface for open_drain mode.
- improve ticks_us() by using the us-counter.
- improve ticks_cpu() by using the CPU's SysTick.

robert-hh committed:
samd: Extend systick_ms to 64 bit.

By adding a 32 bit overflow counter. Allowing to use it for the time
functions.

robert-hh committed:
samd: ADD OPEN_DRAIN mode for pins.

- Use mphal_xx functions whenever possible.
- Remove obsolete includes.
- Clean up traces of a non-functional pin.irq() from earlier builds.
  Pin.irq() will be added in further commits in a working manner.

robert-hh committed:
samd: Allow to specify a pin or led by it's name as a string.

The names are defined in pins.csv.

robert-hh committed:
samd: Enable SoftSPI and SoftI2C.

robert-hh commented:
So I started to add a few more commits along the steps sketched above. That exercise of pulling the code apart and sorting it again actually improves it.

robert-hh committed:
samd: Create pin_af_table.c from the pin_af_table_SAMDxx.csv files.

This step just creates the table.
The firmware cannot be built at this commit step. The next commit will
complete the pin-af mechanism.

robert-hh committed:
samd: Add the pin af table and it's helper functions.

The pin af table is a representation of the MUX table from the data
sheet. It provides information for each pin about the supported device
functions. That information is needed by pin.irq, machine.ADC,
machine.PWM, machine.UART, machine.SPI and machine.I2C. For each
of these, the table tells for each pin, which device number, af number
and pad number is assigned. Using the table gives a straight uniform
access to the information, where the benefit outweights the size of the
table, which is not that large.
The tables are MCU-specific. It is not required to tell for each board,
which and where each of the above devices is available. That makes
addding of boards easy.
Note: The information for DAC and I2S was not included, since it affects
only a few pins.

robert-hh committed:
samd: Add the machine.ADC class.

With the method read_u16(). Keyowrd argumemts of the constructor:

bits = n  The resolution. Default is 12
average=n The average of samples, which are taken and cumulated. The
          default value is 16. Averaging by hw is faster than
          averaging in code.

The ADC runs at a clock freq 1.5 MHz. A single 12 bit conversion takes
8 µs.

robert-hh committed:
samd: Add the machine.PWM class.

- 3 to 5 different frequency groups.
- Freq range 1Hz - 24 MHz.
- Duty rate stays stable on freq change.
- Keyword options to the PWM constructor:

+ device=n select a specific PWM device. If no device is specified,
            a free device is chosen, if available at that pin.
+ freq=nnnn
+ duty_u16=nnnn
+ duty_ns=nnnn
+ invert=True/False Allowing two outputs on the same device/channel to
                    have complemantary signals.

If both freq and duty are provided, PWM output will start immediately.

Pins at the same device have the same frequency. If the PWM output
number exceeds the number of channels at the PWM device, the effctive
channel_no is output_no % channel_count. So with a channel count of 4,
output 7 is assigned to channel 3. Pins at a certain channel have the
same frequency and duty rate, but may be seperately inverted.

robert-hh committed:
samd: Rework the samd_isr.c.

- Have two separate tables for SAM21 and SAMD51.
- Use a short table for SAMD21.
- Add a comment to each line telling what it's for, making further
  use easier.
- Add preliminary handlers/entries for PendSV, EIC and Sercom.
  These will be replaced later when the respecitve modules are added.

robert-hh committed:
samd: Add pin.irq() to the machine.pin class.

Confirming to the docs. There are 16 IRQ channels available, which will
be used as assignable to the GPIO numbers. In most cases, the irq
channel is GPIO_no % 16.

robert-hh committed:
samd: Add disable_irq, enable_irq() and idle() to machine.

No specific features.

robert-hh committed:
samd: Add the class machine.UART.

All board pins that have UART's assigned can be used. Baud rate range
is 75 Baud to ~2 MBaud.

No flow control yet, and only RX is buffered. Tx buffer and flow
control may be added later for SAMD51 with it's larger RAM and Flash.

robert-hh committed:
samd: Add the class machine.SPI.

Suported by both SAMD21 and SAMD51. It follows the generic API, except
for the bits=nn option, which is not implemented (yet).

robert-hh committed:
samd: Add the class machine.I2C.

Using the common API. Tested with SAMD21 and SAMD51 boards.

robert-hh committed:
samd: Add uos.dupterm().

robert-hh committed:
samd: Add uos.urandom() for SAMD51.

Based on the hardware RNG.

robert-hh committed:
samd: Use the common execution mode of boot.py and main.py.

- Do not execute main.py is boot.py failed.
- On a forced exit, do a soft reset.

robert-hh committed:
samd: Enable a few more MicroPython features.

- execute .mpy files
- allow const()
- enable auto-indent in REPL
- enable enumerate, min/max, attrtuple, input

robert-hh committed:
samd: Change the way a Sercom ISR is registered and called.

Code size diff: +12 Bytes
BSS diff: -12 Bytes
RAM usage: +16 Bytes
Speed increase: a few clock cycles per call
Style improvement: ++

robert-hh committed:
samd: Add machine.timer based on the shared SoftTimer.

robert-hh committed:
samd: Add the machine.WDT class.

robert-hh committed:
samd: Add a method pininfo() to the module samd.

pininfo() returns the data stored in the pin af table for a pin.
Using a small script, a nice representation of the table can be created.

robert-hh committed:
samd: Shrink the machine.LED class.

By reducing the methods to on(), off(), toggle() and call, and
using the method implementation of the machine.Pin class.

The code size reduction is 756 byte.

robert-hh commented:
@dpgeorge  This PR is the resorted version of PR #8533. The sequence of commits is slightly change compared to my comment above. The size for a SAMD21 build is now 152k. At an available flash size of 192k, there is now 40 k space left. I prefer to leave ~20k for frozen bytecode or as headroom for source code level debugging of single files. That means, there is about 20k left for code/data. Some package size estimations:
- floating point support: 12k
- Asyncio: 1.5k
- Unicode: 2k

Classes and function on my list:
- DAC (planned next to match ADC): 1k - 2 k, depending on the functionality
- time_pulses (should be minor in size, since the prerequisites are present)
- Onewire 

I tried to implement bitstream, but the pulse time accuracy was bad.
And there is still the bug in TinyUSB/USB.

robert-hh committed:
samd: Add missing/lost board definitions.

- pin definitions for ADAFRUIT_FEATHER_Mx_EXPRESS and
  ADAFRUIT_ITSYBITSY_M4_EXPRESS.
- For ADAFRUIT_ITSYBITSY_M0_EXPRESS, change the MISO/MOSI name.
- For MINISAM_M4, add the default SPI pins.
- For boards with 32k crystal add the XOSC32K setting.

robert-hh committed:
samd: Add the class machine.DAC.

1 Channel @ 10 bit for SAMD21, 2 channels @ 12 bit for SAMD51.
Instantiation by:

dac = machine.DAC(ch) # 0 or 1

Method write:

dac.write(value)

The output voltage range is 0..Vdd.

robert-hh committed:
samd: Add machine.time_pulse_us.

Software based. Resolution:

+/- 2µs on SAMD51
+/- 4µs on SAMD21

robert-hh committed:
samd: Add a variants policy for SAMD21 and SAMD51.

Which contains a mpconfigmcu.h, mpconfigmcu.mk and manifest.py file
for each MCU group. That looks better than the previous choice.

Note: I did not change the earlier commit, because that would have
cause too much trouble in the commit series.

robert-hh committed:
samd: Add a mp_hal_ticks_ms_64() function.

Returning a 64 bit number. This will be used by the utime module and
the machine.UART module for timeout avoiding overflow.

robert-hh committed:
samd: Support buffered TX for UART.

It can be enabled/disabled by a configuration switch. The code size
increase is 308 bytes, but it requires RAM space for buffers, the
larger UART object and root pointers.

robert-hh committed:
samd: Enable time.time() based on systick_ms().

Allowing to set a time and retrieve the time. It is based on
systick_ms() with the precision of the mcu clock. Unless that is
based on a crystal, the error seen was about 0.5% at room temperature.

robert-hh committed:
samd: Use CYCCNT for SAMD51's mp_hal_ticks_cpu().

And use mp_hal_ticks_us() for SAM21's mp_hal_ticks_cpu(). The
SAMD21 has no CYCCNT register, and the SysTick register has only
a 1 ms span (== 48000 count range).

robert-hh committed:
samd: Add the function machine.bitstream().

The SAMD21 implementation is an adaption of @jimmo's code for STM32Lxx.
The only changes are the addresses and names of the port registers
and the timing parameters.

SAMD21:
The precision is about +/-25ns at 48MHz clock frequency.The first two
cycles are about 40-60 ns longer than set.  But still good enough to
drive a neopixel device.

SAMD51:
The precision is about +/-30ns at 120MHz clock frequency. Good enough
to drive a neopixel device.

robert-hh committed:
samd: Restructure the mpconfigxxx.h files.

- Set the feature level for each MCU. CORE features for SAMD21, and
  EXTRA features for SAMD51.
- Remove all definitions that are included in the core feature level.
- Keep the default settings for feature level and float,
  to make the choice obvious.

robert-hh committed:
same: Split clock_config.c in SAMD21 and SAMD51 files.

And put the into the variants tree. The file got a little bit long
and hard to read.

robert-hh committed:
samd: Two extension to the clock set-up.

Two new compile flags:

MICROPY_HW_DFLL_USB_SYNC

Effective only if DFLL48 does not run from the crystal. It will
synchronize the DFLL48M clock with the USB's SOF pulse. If no USB
is connected, it will fall back to open loop mode. The DFLL48M clock
is then pretty precise, but with a higher clock jitter at SAMD51
devices.

MICROPY_HW_MCU_OSC32KULP

Effective only if the devics uses a crystal as clock source.
Run the MCU clock from the ULP 32kHz oszillator instead of the crystal.
This flag was added to cater for a interference problem
of the crystal and Neopixel/Debug pins at Adafruit FEATHER Mx boards,
which causes the board to crash. Drawback: ticks_ms() and time.time()
vs. than ticks_us() and the peripherals like PWM run at not
synchronous clocks.

robert-hh committed:
samd: Simplify the pin-af-table handling.

- The pin-af-table-SAMDxx.csv file are moved to the variants
  directories with the name as pin-af-table.csv.
- The handling in Makefile and pin_af.c is simplified.

robert-hh commented:
Just rebased to Version 1.19.xx

robert-hh commented:
Four commits: Three of them a just making a cleaner file structure for SAMD21 vs SAMD51 boards. The 4th is a bypass for a PCB design issue of Adafruit Feather M0/M4 boards, which prevented using the on-board crystal for the MCU clock.

robert-hh committed:
samd: Add floating point suport for SAMD21 devices.

For consistency it should be there.

robert-hh committed:
samd: Remove obsolete printf() from the samd module.

Must be a remainder from the early test phase.

robert-hh committed:
samd: Add the math module for SAMD51.

robert-hh committed:
samd: Fixing a dependency problem with "make -j1".

The build directory was not created before attempting to create the
generated files in it.

robert-hh committed:
samd: Fix USB endpoint handling ignoring Ctrl-C.

Porting PR #8040 of @hoihu for SAMD, following the commit
587339022689187a1acbccc1d0e2425a67385ff7.

One small addition: Before executing keyboard interrupt, the input
buffer is cleared.

robert-hh committed:
samd: Change the symbol names for the peripheral clock.

From APB_FREQ to DFLL48M_FREQ
From apb_freq to peripheral_freq

robert-hh committed:
samd: Move the FLASH filesystem definitions to the loader files.

They used to be in mpconfigmcu.h, but have to be different for
different chip variants, like the SAMD51x20.

robert-hh committed:
samd: Use the same loader file for all samd51x19 variants.

robert-hh commented:
Two fixes for old issues, adding float for SAMD21 and math for SAMD51 and some straightening of the code base.

robert-hh committed:
samd: Use lf2s for SAMD51 and lfs1 for SAMD21.

Using lfs1 gives a smaller code.

robert-hh committed:
samd: Initialize readline on boot.

Somehow that was forgotten.

robert-hh commented:
> One thing I noticed was that softtimer.[ch] were copied from stm32. Would be good to first factor that code out into shared/runtime/.

I can do so. Should I change that for STM32 then as well?

Another note: I have a few commits for SAMD, which I did not push yet:

031f711c3 samd: Enable FATFS support for SAMD51.
cf82fed0f samd: Remove a few obsolete lines from Makefile.
a460985e9 samd: Add board files for Sparkfun SAMD51 Thing Plus.
f43437736 samd: Support machine.freq() for SAMD21 as well.
be5917155 samd: Extend the range of machine.freq() for SAMD51.
452751899 samd: Add machine.reset_cause().
67a880210 samd: Add the support class for onewire devices (SAMD51).
e965d52af samd: Add the DHTxx driver for SAMD51 devices (SAMD51).

Shall I add them to the current PR or wait? 
The clock extension widens the range down to 1 MHz, at which rate the current consumption is about 1 mA. Down to 8Mhz, USB still works, at about 2-3 mA.

dpgeorge commented:
> I can do so. Should I change that for STM32 then as well?

If you can make another PR just to factor out the softtimer code from stm32 and renesas-ra, and put it in `shared/runtime`/, that would be great.

> Shall I add them to the current PR or wait?

Please wait!  There is already a lot to review here, and I don't want to hold it up much longer.

robert-hh commented:
OK, I'll wait. I had put these commits intentionally aside.

softtimer is not identical for STM32 and SAMD. The differences are small, just the name of the global timer tick variable and one include. But I can start with a PR for refactoring from STM32 and Renesas-RA.

dpgeorge commented:
> softtimer is not identical for STM32 and SAMD. The differences are small, just the name of the global timer tick variable and one include

Then these items should be made configurable with a `#define` (or similar).

robert-hh commented:
I can put a single #define into the SAMD's mphalport.h

robert-hh committed:
samd: Deinit all timers on soft reset.

That was missing.

ricksorensen commented:
@robert-hh I am not very familiar with git usage in the project domain.  In particular I am not sure of doing a PR on a PR on a PR..  but I would like to be able to help if I can.  I have some SAMD21 devices ((SEEED XIAO, Trinkey M0) and I use the UARTS and pin IO for 1-Wire, DHTs.  I use I2C for RTCs and OLEDs (and LCDs), but I don't have any SPI devices.

I have been working with the samd_extend branch a bit, particularly the machine_i2c implementation.  I have just small devices (SEEED XIAO, Trinkey M0). I tried the `machine.I2C` with a 128x64 OLED and the `ssd1306.py` display driver.  I added in the `framebuf` module (`MICROPY_PY_FRAMEBUF=1`) and recompiled.    I then ran into a timeout error when sending the buffer- it worked fine with the SoftI2C, and a smaller 128x32 OLED worked fine also.   The larger OLED has a 1024 byte buffer.  

I tracked the issue down to the timeout processing in the `machine_i2c_transfer_single` in the loop which allows context switching.    The timeout occured after about 950 bytes had been sent (via logic analyzer).  The actual number varies between tests but is in that ball park.  Increasing the timeout value from 1000 to 2000 allows the buffer transfer to complete, but I think a more robust fix is to reset the timeout in the `common_i2c_irq_handler`  (I added a timeout value to the state object).   
```
#define I2C_TIMEOUT_SEND  (10)     // Added RJS
...
typedef struct _machine_i2c_obj_t {
    mp_obj_base_t base;
    Sercom *instance;
    uint8_t id;
    uint8_t scl;
    uint8_t sda;
    uint8_t state;
    uint32_t freq;
    uint32_t timeout;                    // added RJS
    size_t len;
    uint8_t *buf;
} machine_i2c_obj_t;
....
void common_i2c_irq_handler(int i2c_id) {
....
        } else if (IRQ_DATA_SENT) {
            if (NACK_RECVD) { // e.g. NACK after adress for both read and write.
                self->state = state_nack; // force stop of transmission
                i2c->I2CM.INTFLAG.reg |= SERCOM_I2CM_INTFLAG_MB;
            } else if (self->len > 0) { // data to be sent
                i2c->I2CM.DATA.bit.DATA = *(self->buf)++;
                self->len--;
		self->timeout = I2C_TIMEOUT_SEND;
            } else { // No data left, if there was any.
                self->state = state_done;
                i2c->I2CM.INTFLAG.reg |= SERCOM_I2CM_INTFLAG_MB;
            }
        }
...
// in STATIC int machine_i2c_transfer_single(...) replace
//    timeout with self->timeout
```
Using this I set the I2C_TIMEOUT_SEND to 20 and have seen no problems yet.
If there is anything I can do to help let me know, and/or let me know the process for working with you PRs.

BTW PR in my background i 'problem report' not 'pull request' - that was one of the longer learning curves for me to hurdle (wink)

robert-hh commented:
Thanks for the report and the detailed analysis. It's the right way of doing it here. I'll change that. The timeout was only set as a kind of fallback, such that the code will return at some time. The value was chosen arbitrarily.

robert-hh commented:
@ricksorensen machine_i2c.c is updated at https://github.com/robert-hh/micropython/tree/samd_xtend. For my sh1106 display, I do not see a difference between the two code variants. But maybe you could test the updated version with your set-up.

ricksorensen commented:
@robert-hh Looks good - I cloned a fresh repository, switched to `samd_xtend`, recompiled with `framebuf` and ran the HEAD version with ssd1306 and 128x64 OLED (see XIAO expansion board for specs) - with no timeout error.  
I then went back 1 version in the repository and the timeout occured:
```
Traceback (most recent call last):
  File "<stdin>", line 22, in <module>
  File "/lib/ssd1306.py", line 110, in __init__
  File "/lib/ssd1306.py", line 36, in __init__
  File "/lib/ssd1306.py", line 73, in init_display
  File "/lib/ssd1306.py", line 101, in show
  File "/lib/ssd1306.py", line 119, in write_data
OSError: 116
```
The error actually occurs during the object creation (`oled = SSD1306_I2C(wd, ht, i2c, addr=60)`) during the clearing of the screen.

robert-hh commented:
Thank you for the confirmation. The difference between the SSD1306 driver and the SH1106 driver is, that the SSD1306 drive writes the full buffer in one call, while the Sh1106 driver writes only one page per call.

andypiper commented:
@robert-hh assuming that this PR is the same that I've been testing from your forum thread - this is looking good on the MINI SAM M4 build w/ Thonny for me!

robert-hh commented:
@andypiper Thank you for the feedback.
The PR is similar. At my repository there are two branches. One of them (samd_extend) is linked to the PR, the other (samd_xtend) has a few more commits, but samd_extend is a subset of samd_xtend.

robert-hh commented:
Sorry. I had to add another small commit to deal with commit `88864587f   py/objstr: Always ensure mp_obj_str_from_vstr is unicode-safe.`

robert-hh commented:
This last commit is again just to keep up with changes in the main branch, this time for manifest.py. I attempt to keep that PR on track with the master, so that it can easily be merged if that will ever happen.

robert-hh commented:
@dpgeorge I'm a little bit unsure how to proceed with the SAMD port. This actual PR freezes a state which by the time got an intermediate one. There a a few more commits in a second branch, where I continued the development. List of additional commits below. My concern is, that if you eventually review this PR, there will be changes that are hard to sqeeze in between. So since the first set of commits is mostly identical in both branches, I could as well change the PR for the more complete variant.

List of additional commits in chronological order
```
 samd: Add depencies in Makefile for pins.h.
*samd: Add machine.reset_cause().
+samd: Extend the range of machine.freq().
+samd: Support machine.freq() for SAMD21 as well.
*samd: Add board files for Sparkfun SAMD51 Thing Plus.
 samd: Remove a few obsolete lines from Makefile.
*samd: Enable FAT support for SAMD51.
 samd: Add only existing pins pin_table_af.c.
*samd: Enable os.uname().
 samd: Reset I2C timeout for every transferred byte.
*samd: Implement machine.lightsleep().
 samd: Move SysTick_Config() into the function set_cpu_freq().
 samd: Get the bootloader magic address from the lib.
 samd: Support SoftSPI maximum baud rate mode.
 samd: Document the #defines use in clock_init().
 samd: Two small fixes to machine.uart.
*samd: Implement uart.txdone() and uart.flush().
 samd: Add a DMA manager.
 samd: Add a tc_manager function set.
*samd: Add the dac.write_timed() and dac.deinit() methods.
 samd: Add management code for DMA IRQ handlers.
+samd: Add an optional count argument to dac.write_timed().
*samd: Add a adc.read_timed_into() method.
*samd: Add the math lib support for SAMD21.
+samd: Add a vref=num option to machin.adc and machine.dac.
+samd: Add a callback keyword option to machine.DAC().
+samd: Add a callback keyword option to machine.ADC().
 samd: Squeeze a few bytes out of the code.
*samd: Add the machine.RTC module.
*samd: Allow pin names/numbers instead of pin objects. 

* New feature visible in the API
+ Extension of that feature or follow up for other ports
```

dpgeorge commented:
Sorry I haven't had a chance to look at this again. I still prefer to take small steps, but it seems your steps are faster than what I can review!

Anyway, let me take a look at this PR to see how we can proceed.

robert-hh commented:
Thank you for the swift reply.
About the state of the SAMD port: feature-wise it seems to be pretty complete now compared with other non-networked non-BT ports. So the port development will change from a growing phase into a maintenance phase, with fewer adaptions sometimes.

ricksorensen commented:
@robert-hh I added an issue to the samd_xtend port for PWM processing.  When a PWM.deinit() is called, the TCC clock is disable, and if another PWM which uses that TCC is started the prior PWM is started also- the peripheral mux was not cleared.  The issue has pictures and some suggestions and I am certain some clean up is needed on the clearing/setting of PMUX and/or TCC clocks.

robert-hh commented:
Thanks. I'll look into it. It may need to consider other channels using that tcc as well.

robert-hh commented:
Clearing the Pin Mux is the easy part. But thinking about it I wonder, if deinit should stop the tcc, if other channels are still used. These would be stopped as well.

robert-hh committed:
samd: Some rework of the PWM module.

- Clear the Pin AF mode when calling pwm.deinit().
- Defer setting the Pin AF mode until the freq is set.
- Disable a PWM device only if none of it's outputs are used.
- Store the PWM duty in a separate vector. That way, they
  are available at a frequency change for keeping the duty rates.
- Use the double buffered registers for a glitch free freq change.
- Make freq the second (optional) argument of the constructor, not
  requiring a keyword (keyword can still be used).

dpgeorge commented:
I've done a review of this PR.  I have not looked at every line of code but in general it looks very good.

- You add softtimer.c then remove it in a later commit.  Is it possible to not do this, ie just use the shared one from the beginning?
- Trying it out on the SEEED_XIAO, LED(1) and LED(2) are inverted, the LED turns off when you call `.on()` and on when you call `.off()` (maybe that bug was there before this PR though).
- Printing a Pin object gives something like `GPIO PA17`.  Would be good to make that match other ports, eg print `Pin(...)`.
- `machine.DAC(...)` requires an integer (not a pin), but `machine.ADC(...)` requires a pin.  Would it make sense to requre/allow DAC to take a pin, or is the DAC fixed to a specific pin?  (At least esp32 allows passing in a pin to the constructor.)
- A new `samd.pininfo()` function is added.  I think we should leave this out for now.  We can add something like it later but try to make it general across all ports.

dpgeorge commented on ports/samd/pin_af.c:
```diff
@@ -0,0 +1,125 @@
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * This file provides and checks pin capabilities as required
+ * for USART, I2C, SPI, PWM, ADC
```
This text should go after the license.

robert-hh replied:
OK

dpgeorge commented on ports/samd/pin_af.h:
```diff
@@ -0,0 +1,91 @@
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * This file initialises provides and checks pin capabilities as required
+ * for USART, I2C, SPI, PWM, ADC
```
This text should go after the license.

robert-hh replied:
OK

dpgeorge commented on ports/samd/variants/samd21/mpconfigmcu.h:
```diff
@@ -0,0 +1,36 @@
+
+// Use VfsLfs's types for fileio/textio
+#define mp_type_fileio mp_type_vfs_lfs1_fileio
+#define mp_type_textio mp_type_vfs_lfs1_textio
```
These two defines are no longer needed, they can be removed.

robert-hh replied:
OK

dpgeorge commented on ports/samd/variants/samd51/mpconfigmcu.h:
```diff
@@ -0,0 +1,53 @@
+
+// Use VfsLfs2's types for fileio/textio
+#define mp_type_fileio mp_type_vfs_lfs2_fileio
+#define mp_type_textio mp_type_vfs_lfs2_textio
```
These two defines can be removed.

robert-hh replied:
OK.

dpgeorge commented:
I see you added a variants directory for the different MCU configuration.  The name "variants" clashes with the concept of variants in the unix port, and now board variants in other ports (stm32, rp2).  So please rename this directory to something else, I suggest `mcu`, or ever `mcu_series`.

dpgeorge commented on ports/samd/boards/ADAFRUIT_ITSYBITSY_M0_EXPRESS/pins.csv:
```diff
@@ -0,0 +1,40 @@
+PIN_PA15,D5
+-
+PIN_PA21,D7
+-
```
I don't understand these `-` entries, what do they do?  Isn't it simpler to just leave these lines out altogether?

robert-hh replied:
For Pins, numbers can be used, which are the item number in the pins.csv file. Since there is no D6, a dummy definition line is used. That scheme existed before. I have a newer (not committed) variant, where the pin number just reflects the Port number, so PA12 is pin 12, PB3 is pin 35.,.... That allows accessing pins not defined in pins.csv. But I keep that scheme for the end.

dpgeorge replied:
I see.  Let's leave it then and use your new scheme later on.

dpgeorge commented on ports/samd/boards/samd51x19a.ld:
```diff
@@ -13,5 +13,8 @@ MEMORY
 _sstack = _estack - 16K;
 
+_oflash_fs = ORIGIN(FLASH) + 384K - 16K;
+_sflash_fs = LENGTH(FLASH) - 384K + 16K - 1;
```
IMO these names are confusing, because "s" as a prefix usually means "start" in linker variables.  I suggest following stm32 and using `_micropy_hw_internal_flash_storage_start` and `_micropy_hw_internal_flash_storage_end` (or `_micropy_hw_internal_flash_storage_length`).

robert-hh replied:
Thanks for starting the review.
These names were existing before (committed  2019) and I did not change these. Obviously everything can be changed.

dpgeorge replied:
Ah, OK.  In that case leave them as-is, we can change them later if needed.

robert-hh commented:
>     You add softtimer.c then remove it in a later commit. Is it possible to not do this, ie just use the shared one from the beginning?

Re-odering the commits is difficult. It all relates to the long time the PR was open and my attempt to keep it compatible with a moving target. The shared/softimer.c change was made when that was introduced. So the commit series mirrors partly the history of micropython.

>     Trying it out on the SEEED_XIAO, LED(1) and LED(2) are inverted, the LED turns off when you call .on() and on when you call .off() (maybe that bug was there before this PR though).

The LED class is just a subset of the PIN class. So it depends on the wiring of the PED whetehr is goes on of off with a LED.on() call. Changing that would require a "inverted" flag at the LED definitions. I'd prefer to drop the LED class.

>    Printing a Pin object gives something like GPIO PA17. Would be good to make that match other ports, eg print Pin(...).

Can be done. Would you like to see more information than the port number?

>    machine.DAC(...) requires an integer (not a pin), but machine.ADC(...) requires a pin. Would it make sense to requre/allow DAC to take a pin, or is the DAC fixed to a specific pin? (At least esp32 allows passing in a pin to the constructor.)

DAC is fixed to specific pins. For SAMD21 it's PA02, for SAMD51 it's PA02 and PA05

>     A new samd.pininfo() function is added. I think we should leave this out for now. We can add something like it later but try to make it general across all ports.

That call is very helpful to tell, which I/O function is available at a pin. I used it to create the Pin capabilities table in the documentation.

robert-hh commented:
> I see you added a variants directory for the different MCU configuration. The name "variants" clashes with the concept of variants in the unix port, and now board variants in other ports (stm32, rp2). So please rename this directory to something else, I suggest mcu, or even mcu_series.

`mcu` was one of the names which I considered. And the change is easy, just in Makefile.

robert-hh committed:
samd: Text changes to pin_af.*, samd_isr.c and mpconfigmcu.h.

robert-hh committed:
samd: Change the directory name 'variants' to 'mcu'.

robert-hh committed:
samd: Change the pin.print() and led.print() output.

It now prints lines like:

Pin("D9", mode=IN, pull=PULL_UP, GPIO=PA07)

or

LED("LED")

showing for consistency the names as given in pins.csv. For pins,
the GPIO numer is printed as well as reference.

robert-hh commented:
> You add softtimer.c then remove it in a later commit. Is it possible to not do this, ie just use the shared one from the beginning?

So I moved using the shared softtimer to the beginning. Since that changed the commit sequence, you have to re-load the PR.

robert-hh commented:
@dpgeorge I do not know how you want to proceed with this PR. Alternatives:

a) Merge this PR withe the requested changes.
b) Look at the second batch of commits first and then Merge it.

In case of a) there are two commits in the seconds batch, which should be included, because they fix bugs to machine.uart and machine.pwm. The change to machine.uart is minor and does only affect a configuration with non-buffered transmit. The bug in pwm affects the duty rate at frequency changes, where all channels of a PWM device are set to the same duty rate on a frequency change. There are other changes too in the PWM module for a more clean and compatible performance.

mattytrentini commented:
Thanks @robert-hh - and @dpgeorge for the review - there's a _lot_ of work in this PR and it will be a _very_ welcome addition.

dpgeorge commented on ports/samd/samd_isr.c:
```diff
@@ -1,20 +1,20 @@
 /*
  * This file is part of the MicroPython project, http://micropython.org/
- *
+
```
something went wrong here, please don't change this license header

dpgeorge commented:
> Can be done. Would you like to see more information than the port number?

Would be good to provide the mode and pull state, eg:
```
Pin(0, mode=OUT, pull=PULL_UP)
```

dpgeorge commented:
> In case of a) there are two commits in the seconds batch, which should be included, because they fix bugs to machine.uart and machine.pwm.

I'm happy to merge this PR with the changes you've made (plus my two new comments above).  Those bug fixes can wait until the second batch.

robert-hh commented:
@dpgeorge The changes are made as requested by updating the previous commits. But I could not stand it leaving PWM as it was. Known inconsistencies should be fixed. So I added the PWM fix. The UART fix is not criticle, because it does not show up in the actual configuration with transmit buffer.

dpgeorge commented:
Thanks for updating.  I should be able to merge this shortly.  I may do some squashing of some commits during merge.

robert-hh commented:
Thanks. No problem about squashing. There are quite a few pretty small commits.

robert-hh commented:
The matching update for the documentation is in PR #8798.

dpgeorge commented:
Merged in 85afed569d3d435b2611856356ff9c78d244f25c through 366c801b3553191241e329c2f6d7f718f49af105

Thank you very much @robert-hh.  After going through the commits in detail, it's obvious that a LOT of work went into all of this.  It's a really great addition, it makes the samd port close to full-featured.

-----

I squashed a few of the later fix-up commits into the earlier ones.  But I only made a very few minor changes overall.  For your interest to see what I did change, this is the diff (the changes I made to this PR before merging):
```diff
diff --git a/ports/samd/Makefile b/ports/samd/Makefile
index 7510066cf..5ff00f21f 100644
--- a/ports/samd/Makefile
+++ b/ports/samd/Makefile
@@ -90,7 +90,7 @@ SRC_C += \
        mcu/$(MCU_SERIES_LOWER)/clock_config.c \
        help.c \
        machine_adc.c \
-       machine_bitstream.c \
+       machine_bitstream.c \
        machine_dac.c \
        machine_i2c.c \
        machine_led.c \
diff --git a/ports/samd/boards/ADAFRUIT_ITSYBITSY_M4_EXPRESS/pins.csv b/ports/samd/boards/ADAFRUIT_ITSYBITSY_M4_EXPRESS/pins.csv
index 3ad6c4395..ce760a269 100644
--- a/ports/samd/boards/ADAFRUIT_ITSYBITSY_M4_EXPRESS/pins.csv
+++ b/ports/samd/boards/ADAFRUIT_ITSYBITSY_M4_EXPRESS/pins.csv
@@ -13,7 +13,7 @@ PIN_PA14,D4
 PIN_PA15,D5
 -
 PIN_PA18,D7
-- 
+-
 PIN_PA19,D9
 PIN_PA20,D10
 PIN_PA21,D11
diff --git a/ports/samd/boards/SAMD21_XPLAINED_PRO/pins.csv b/ports/samd/boards/SAMD21_XPLAINED_PRO/pins.csv
index 85f76381c..438119d90 100644
--- a/ports/samd/boards/SAMD21_XPLAINED_PRO/pins.csv
+++ b/ports/samd/boards/SAMD21_XPLAINED_PRO/pins.csv
@@ -33,7 +33,7 @@ PIN_PB13,EXT2_PIN8
 PIN_PB14,EXT2_PIN9
 PIN_PB15,EXT2_PIN10
 -
-- 
+-
 PIN_PB11,EXT2_PIN13
 PIN_PB10,EXT2_PIN14
 PIN_PA17,EXT2_PIN15
diff --git a/ports/samd/main.c b/ports/samd/main.c
index 382e7702e..0f24e2382 100644
--- a/ports/samd/main.c
+++ b/ports/samd/main.c
@@ -29,10 +29,10 @@
 #include "py/gc.h"
 #include "py/mperrno.h"
 #include "py/stackctrl.h"
+#include "shared/readline/readline.h"
 #include "shared/runtime/gchelper.h"
 #include "shared/runtime/pyexec.h"
 #include "shared/runtime/softtimer.h"
-#include "shared/readline/readline.h"
 
 extern uint8_t _sstack, _estack, _sheap, _eheap;
 extern void adc_deinit_all(void);
@@ -48,6 +48,7 @@ void samd_main(void) {
     for (;;) {
         gc_init(&_sheap, &_eheap);
         mp_init();
+
         // Initialise sub-systems.
         readline_init0();
 
diff --git a/ports/samd/mcu/samd21/mpconfigmcu.mk b/ports/samd/mcu/samd21/mpconfigmcu.mk
index 1d3b568c6..cc435da8c 100644
--- a/ports/samd/mcu/samd21/mpconfigmcu.mk
+++ b/ports/samd/mcu/samd21/mpconfigmcu.mk
@@ -1,2 +1 @@
-
 SRC_S += shared/runtime/gchelper_m0.s
diff --git a/ports/samd/mcu/samd51/mpconfigmcu.mk b/ports/samd/mcu/samd51/mpconfigmcu.mk
index cc5cb1f31..e83e8911d 100644
--- a/ports/samd/mcu/samd51/mpconfigmcu.mk
+++ b/ports/samd/mcu/samd51/mpconfigmcu.mk
@@ -1,4 +1,3 @@
-
 MICROPY_VFS_LFS2 ?= 1
 
 SRC_S += shared/runtime/gchelper_m3.s
diff --git a/ports/samd/modmachine.c b/ports/samd/modmachine.c
index b10771d47..f6cf7f815 100644
--- a/ports/samd/modmachine.c
+++ b/ports/samd/modmachine.c
@@ -168,10 +168,7 @@ STATIC const mp_rom_map_elem_t machine_module_globals_table[] = {
     { MP_ROM_QSTR(MP_QSTR_disable_irq),         MP_ROM_PTR(&machine_disable_irq_obj) },
     { MP_ROM_QSTR(MP_QSTR_enable_irq),          MP_ROM_PTR(&machine_enable_irq_obj) },
     { MP_ROM_QSTR(MP_QSTR_time_pulse_us),       MP_ROM_PTR(&machine_time_pulse_us_obj) },
-
-    // #if defined (MCU_SAMD51)
     { MP_ROM_QSTR(MP_QSTR_bitstream),           MP_ROM_PTR(&machine_bitstream_obj) },
-    // #endif
 };
 STATIC MP_DEFINE_CONST_DICT(machine_module_globals, machine_module_globals_table);
 
diff --git a/ports/samd/moduos.c b/ports/samd/moduos.c
index 0cddbf748..e0237fd88 100644
--- a/ports/samd/moduos.c
+++ b/ports/samd/moduos.c
@@ -38,7 +38,6 @@
 static bool initialized = false;
 
 STATIC void trng_start(void) {
-
     if (!initialized) {
         MCLK->APBCMASK.bit.TRNG_ = 1;
         REG_TRNG_CTRLA = TRNG_CTRLA_ENABLE;
@@ -47,7 +46,6 @@ STATIC void trng_start(void) {
 }
 
 uint32_t trng_random_u32(void) {
-
     trng_start();
     while ((REG_TRNG_INTFLAG & TRNG_INTFLAG_DATARDY) == 0) {
     }
@@ -73,8 +71,8 @@ STATIC mp_obj_t mp_uos_urandom(mp_obj_t num) {
 }
 STATIC MP_DEFINE_CONST_FUN_OBJ_1(mp_uos_urandom_obj, mp_uos_urandom);
 
-#endif  // MICROPY_PY_UOS_URANDOM
-#endif  // defined(MCU_SAMD51)
+#endif // MICROPY_PY_UOS_URANDOM
+#endif // defined(MCU_SAMD51)
 
 #if MICROPY_PY_UOS_DUPTERM_BUILTIN_STREAM
 bool mp_uos_dupterm_is_builtin_stream(mp_const_obj_t stream) {
diff --git a/ports/samd/samd_soc.c b/ports/samd/samd_soc.c
index 729bd356e..f4a27f3df 100644
--- a/ports/samd/samd_soc.c
+++ b/ports/samd/samd_soc.c
@@ -64,7 +64,7 @@ static void usb_init(void) {
     tusb_init();
 }
 
-// Initialize the µs counter on TC 0/1
+// Initialize the microsecond counter on TC 0/1
 void init_us_counter(void) {
     #if defined(MCU_SAMD21)
 ```

robert-hh commented:
Thank you very much. The changes look fine. Now I can proceed with the second set, which is smaller and more maintenance that feature expansion.

robert-hh commented:
I just see that you did some rewriting of the commit messages, which improves them. But you skipped the last commit with the code fixes for PWM. That's a pity. So it will be in the second batch.

robert-hh commented:
Yes, I noticed on a second glance that you went though the commits piece by piece, and improved the language. It looks much better now.
